network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+        )
+        .unwrap(),
+    )
+});
+
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+    // Collect the set of address updates.
+    let updates: Vec<ConnectivityRequest> = validator_keys
+        .into_iter()
+        .map(|validator| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                validator.account_address().clone(),
+                vec![
+                    NetworkAddress::try_from(&validator.config().validator_network_address)
+                        .expect("WTF"),
+                ],
+            )
+        })
+        .collect();
+    // Collect the set of EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        validator_keys
+            .into_iter()
+            .map(|validator| {
+                (
+                    *validator.account_address(),
+                    NetworkPublicKeys {
+                        identity_public_key: validator.network_identity_public_key(),
+                        signing_public_key: validator.network_signing_public_key().clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+    updates
+}
+
+fn extract_full_node_updates(full_node_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let full_node_infos = full_node_set.payload().to_vec();
+
+    // Collect the set of updated addresses.
+    let updates: Vec<ConnectivityRequest> = full_node_infos
+        .into_iter()
+        .map(|full_node| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                // TODO:  this seems wrong for the account address.  Is it shared with the validator?
+                full_node.account_address().clone(),
+                vec![NetworkAddress::try_from(
+                    &full_node.config().full_node_network_address.clone(),
+                )
+                .expect("WTF")],
+            )
+        })
+        .collect();
+
+    // Collect the EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        full_node_infos
+            .into_iter()
+            .map(|full_node| {
+                (
+                    // TODO:  This seems like it is wrong and should be something else.  Or is the full_node account address the same as the validator?
+                    full_node.account_address().clone(),
+                    NetworkPublicKeys {
+                        identity_public_key: full_node
+                            .config()
+                            .full_node_network_identity_public_key
+                            .clone(),
+                        // TODO:  there are not separate keys for the full_node.  Are they therefore the same?
```

Comment:
These would be the same... our current approach is that a validator also runs a full node and that these are tightly coupled. Hence these are supposed to be trusted fullnodes that should map to a specific validator. Longer term, we might envision supporting 3rd party fullnodes on chain, but that is a bit more of utility / complexity that hasn't been properly evaluated.

Reply:
ahh, so this should be the validator public key?